vpn-router-web/src/main/java/vpnrouter/web/ui/clients/CompetitionListenerImpl.java

```diff
@@ -0,0 +1,56 @@
+package vpnrouter.web.ui.clients;
+
+import com.vaadin.flow.component.UI;
+import com.vaadin.flow.component.notification.Notification;
+import lombok.RequiredArgsConstructor;
+import vpnrouter.api.client.ClientDetectionService;
+
+import java.util.Timer;
+import java.util.TimerTask;
+
+@RequiredArgsConstructor
+public class CompetitionListenerImpl implements ClientDetectionService.CompletionListener {
+    private final int DELAY_MILLISECONDS = 1000;
+    private final UI ui;
+
+    @Override
+    public void onStart() {
+        ui.access(() -> Notification.show("New detection has started"));
+    }
+
+    @Override
+    public void onAlreadyRunning() {
+        ui.access(() -> Notification.show("Detection is already in progress"));
+    }
+
+    @Override
+    public void onNewClientsNotFound() {
+        ui.access(() -> Notification.show("Detection completed: no new clients found"));
+    }
+
+    @Override
+    public void onNewClientsFound(int newClientsCount) {
+        ui.access(() -> {
+                    Notification.show(
+                            "Detection completed: %s new clients found".formatted(newClientsCount),
+                            DELAY_MILLISECONDS,
+                            Notification.Position.BOTTOM_START
+                    );
+                    new Timer().schedule(
+                            new TimerTask() {
+                                @Override
+                                public void run() {
+                                    ui.access(() -> ui.getPage().reload());
```

Comment:
Page reloading is too rough method to refresh data -- should be a bit smarter/optimal